tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -58,84 +62,86 @@ def jax_lanczos(matvec: Callable, arguments: List, init: jax.ShapedArray,
         falls below `landelta`, iteration is stopped.
       reortho: If `True`, reorthogonalize all krylov vectors at each step.
         This should be used if `neig>1`.
+      precision: jax.lax.Precision type used in jax.numpy.vdot
     Returns:
       jax.numpy.ndarray: Eigenvalues
       list: Eigenvectors
     """
-
+    shape = init.shape
+    dtype = init.dtype
+    # TODO (mganahl): replace with restarted classical gram-schmidt
     def body_modified_gram_schmidt(i, vals):
       vector, krylov_vectors = vals
       v = krylov_vectors[i, :]
-      vector -= jax.numpy.vdot(
-          v, vector, precision=precision) * jax.numpy.reshape(v, vector.shape)
+      vector = vector - jax.numpy.vdot(v, vector, precision=precision) * v
       return [vector, krylov_vectors]
 
     def body_lanczos(vals):
-      current_vector, krylov_vectors, vector_norms = vals[0:3]
-      diagonal_elements, matvec, args, _ = vals[3:7]
-      threshold, i, maxiteration = vals[7:]
-      norm = jax.numpy.linalg.norm(current_vector)
-      normalized_vector = current_vector / norm
-      normalized_vector, krylov_vectors = jax.lax.cond(
-          reortho, True,
-          lambda x: jax.lax.fori_loop(0, i, body_modified_gram_schmidt,
-                                      [normalized_vector, krylov_vectors]),
-          False, lambda x: [normalized_vector, krylov_vectors])
-      Av = matvec(normalized_vector, *args)
-
-      diag_element = jax.numpy.vdot(normalized_vector, Av, precision=precision)
-
-      res = jax.numpy.reshape(
-          jax.numpy.ravel(Av) -
-          jax.numpy.ravel(normalized_vector) * diag_element -
-          krylov_vectors[i - 1] * norm, Av.shape)
-      krylov_vectors = jax.ops.index_update(krylov_vectors, jax.ops.index[i, :],
-                                            jax.numpy.ravel(normalized_vector))
-
-      vector_norms = jax.ops.index_update(vector_norms, jax.ops.index[i - 1],
-                                          norm)
-      diagonal_elements = jax.ops.index_update(diagonal_elements,
-                                               jax.ops.index[i - 1],
-                                               diag_element)
-
-      return [
-          res, krylov_vectors, vector_norms, diagonal_elements, matvec, args,
-          norm, threshold, i + 1, maxiteration
-      ]
+      krylov_vectors, alphas, betas, i = vals
+      previous_vector = krylov_vectors[i, :]
+
+      previous_vector, krylov_vectors = jax.lax.cond(
+          reortho,
+          lambda x: jax.lax.fori_loop(1, i, body_modified_gram_schmidt, x),
```

Comment:
cond runs both branches and returns only the result dictated by the predicate, and thus using it here gives no performance benefit. Is this intended? If you want to actually stop early, I think while_loop is the way to go.

Reply:
This is `cond` statements branches between doing explicit reorthogonalization of krylov vectors or not doing it, in the latter case relying on the (ill-conditioned) assumption that constructed krylov vectors are already orthogonal.

Reply:
Right, but if your goal is to save on computation in the latter case this will not achieve that. If it's merely for correctness it's fine.

Reply:
hmm, why not?

Reply:
cond always executes both branches

Reply:
Is this meant as an optimization? In that case it should probably be changed.

Reply:
I think what you want is to use the cond as the termination condition in a while loop

Reply:
is that also true after `jit`?

Reply:
Yeah I'm pretty sure. Everything gets traced inside one of those control flow thingies anyway.

Reply:
just checked, you're right. I'll fix this then